components/board/genericlinux/i2c.go

```diff
@@ -0,0 +1,126 @@
+package genericlinux
+
+import (
+	"context"
+	"encoding/binary"
+	"sync"
+
+	"github.com/edaniels/golog"
+	"periph.io/x/conn/v3/i2c"
+	"periph.io/x/conn/v3/i2c/i2creg"
+	"periph.io/x/host/v3"
+
+	"go.viam.com/rdk/components/board"
+)
+
+func init() {
+	if _, err := host.Init(); err != nil {
+		golog.Global().Debugw("error initializing host", "error", err)
+	}
+}
+
+type i2cBus struct {
+	internal i2c.BusCloser
+	mu       sync.Mutex
+}
+
+func newI2cBus(deviceName string) (*i2cBus, error) {
+	// We return a pointer to an i2cBus instead of an i2cBus itself so that we can return nil if
+	// something goes wrong.
+	bus, err := i2creg.Open(deviceName)
+	if err != nil {
+		return nil, err
+	}
+	return &i2cBus{internal: bus}, nil
+}
+
+// This lets the i2cBus type implement the board.I2C interface.
+func (bus *i2cBus) OpenHandle(addr byte) (board.I2CHandle, error) {
+	bus.mu.Lock() // Lock the bus so no other handle can use it until this one is closed.
+	return &i2cHandle{device: &i2c.Dev{Bus: bus.internal, Addr: uint16(addr)}, mu: &bus.mu}, nil
+}
+
+type i2cHandle struct {
+	device *i2c.Dev    // Will become nil if we Close() the handle
+	mu     *sync.Mutex // Points to the i2cBus' mutex
+}
+
+// This helps the i2cHandle struct implement the board.I2CHandle interface.
+func (h *i2cHandle) Write(ctx context.Context, tx []byte) error {
+	return h.device.Tx(tx, nil)
+}
+
+// This helps the i2cHandle struct implement the board.I2CHandle interface.
+func (h *i2cHandle) Read(ctx context.Context, count int) ([]byte, error) {
+	buffer := make([]byte, count)
+	err := h.device.Tx(nil, buffer)
+	if err != nil {
+		return nil, err
+	}
+	return buffer, nil
+}
+
+// This is a private helper function, used to implement the rest of the board.I2CHandle interface.
+func (h *i2cHandle) transactAtRegister(register byte, w, r []byte) error {
+	if w == nil {
+		w = []byte{}
+	}
+	fullW := make([]byte, len(w)+1)
+	fullW[0] = register
+	copy(fullW[1:], w)
+	return h.device.Tx(fullW, r)
+}
+
+// This helps the i2cHandle struct implement the board.I2CHandle interface.
+func (h *i2cHandle) ReadByteData(ctx context.Context, register byte) (byte, error) {
+	result := make([]byte, 1)
+	err := h.transactAtRegister(register, nil, result)
+	if err != nil {
+		return 0, err
+	}
+	return result[0], nil
```

Comment:
[q[] this is because this method only handles the MSB? comment could be updated for clarity

Reply:
Things on an I2C device don't really have MSBs and LSBs: each register is 1 byte long, but adjacent ones sometimes contain the MSB and LSB of the same value, and reading the two of them together can be more convenient than reading them separately. This function is for reading a single register, whereas the `*Word*` functions below are for reading pairs of adjacent registers and treating the two bytes as though they're a (MSB, LSB) pair.